File: typing_trainer.cs
// typing_trainer.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text.Json;
using System.Diagnostics;

class TypingTrainer
{
    static string Colorize(string text, string color)
    {
        string col = color switch
        {
            "green" => "\x1b[92m",
            "red" => "\x1b[91m",
            "yellow" => "\x1b[93m",
            "blue" => "\x1b[94m",
            "bold" => "\x1b[1m",
            _ => "\x1b[0m"
        };
        return col + text + "\x1b[0m";
    }

    static Dictionary<string, Dictionary<string, List<string>>> WORDS = new Dictionary<string, Dictionary<string, List<string>>>()
    {
        {"ru", new Dictionary<string, List<string>>()
        {
            {"easy", new List<string>{"кот","нос","рот","дом","лес","море","вода","рука","нога","мир"}},
            {"medium", new List<string>{"компьютер","программа","алгоритм","интернет","сервер","сайт","код","данные"}},
            {"hard", new List<string>{"программирование","искусственный интеллект","коммуникация","информация","технология"}}
        }},
        {"en", new Dictionary<string, List<string>>()
        {
            {"easy", new List<string>{"cat","dog","sun","run","jump","tree","book","fish","bird","hand"}},
            {"medium", new List<string>{"computer","program","algorithm","internet","server","website","code","data"}},
            {"hard", new List<string>{"programming","artificial intelligence","communication","information","technology"}}
        }}
    };

    class Stats
    {
        public double best_wpm { get; set; }
        public int games { get; set; }
        public double total_accuracy { get; set; }
    }

    private string lang;
    private string level;
    private string text;
    private string userInput;
    private DateTime start;
    private DateTime end;
    private Stats stats;
    private string statsFile;

    public TypingTrainer(string lang, string level)
    {
        this.lang = lang;
        this.level = level;
        statsFile = Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.UserProfile), $".typing_trainer_{lang}.json");
        LoadStats();
        GenerateText();
    }

    void LoadStats()
    {
        if (File.Exists(statsFile))
        {
            try
            {
                string json = File.ReadAllText(statsFile);
                stats = JsonSerializer.Deserialize<Stats>(json);
            }
            catch { stats = new Stats(); }
        }
        else stats = new Stats();
    }

    void SaveStats()
    {
        string json = JsonSerializer.Serialize(stats);
        File.WriteAllText(statsFile, json);
    }

    void GenerateText()
    {
        var words = WORDS[lang][level];
        Random rnd = new Random();
        int count = level == "hard" ? rnd.Next(8, 16) : rnd.Next(5, 11);
        var textParts = new List<string>();
        for (int i = 0; i < count; i++)
        {
            if (i > 0)
            {
                if (level == "hard" && rnd.NextDouble() < 0.3)
                    textParts.Add(", ");
                else
                    textParts.Add(" ");
            }
            textParts.Add(words[rnd.Next(words.Count)]);
            if (level == "hard" && (i == count-1 || rnd.NextDouble() < 0.2))
                textParts.Add(". ");
        }
        text = string.Concat(textParts).Trim();
    }

    void DisplayText()
    {
        Console.WriteLine(Colorize("\nНапечатайте следующий текст:", "bold"));
        Console.WriteLine(Colorize(text, "blue"));
        Console.WriteLine("\nНажмите Enter после завершения ввода. Время начнётся с первой буквы.");
        Console.WriteLine("Для выхода нажмите Ctrl+C.\n");
    }

    void GetUserInput()
    {
        start = DateTime.Now;
        userInput = Console.ReadLine();
        end = DateTime.Now;
    }

    void CalculateAndDisplay()
    {
        string original = text;
        string typed = userInput ?? "";
        int total = original.Length;
        int typedLen = typed.Length;
        int correct = 0;
        int errors = 0;
        for (int i = 0; i < total; i++)
        {
            if (i < typedLen && original[i] == typed[i])
                correct++;
            else
                errors++;
        }
        if (typedLen < total) errors += total - typedLen;
        double elapsed = (end - start).TotalSeconds;
        double minutes = elapsed / 60.0;
        double wpm = (correct / 5.0) / minutes;
        double accuracy = (correct / (double)total) * 100.0;

        Console.WriteLine(Colorize("\nРезультат:", "bold"));
        for (int i = 0; i < total; i++)
        {
            if (i < typedLen && original[i] == typed[i])
                Console.Write(Colorize(original[i].ToString(), "green"));
            else
                Console.Write(Colorize(original[i].ToString(), "red"));
        }
        Console.WriteLine();

        Console.WriteLine(Colorize($"Скорость: {wpm:F1} WPM", "yellow"));
        Console.WriteLine($"Точность: {accuracy:F1}%");
        Console.WriteLine($"Правильных символов: {correct}/{total}");
        Console.WriteLine($"Ошибок: {errors}");
        Console.WriteLine($"Время: {elapsed:F2} сек");

        stats.games++;
        if (wpm > stats.best_wpm)
        {
            stats.best_wpm = wpm;
            Console.WriteLine(Colorize("🏆 Новый рекорд!", "green"));
        }
        stats.total_accuracy = (stats.total_accuracy * (stats.games - 1) + accuracy) / stats.games;
        SaveStats();
        Console.WriteLine(Colorize($"Лучший результат: {stats.best_wpm:F1} WPM", "blue"));
    }

    public void Run()
    {
        Console.WriteLine(Colorize("⌨️  Тренажёр слепой печати", "bold"));
        Console.WriteLine($"Язык: {lang}, Уровень: {level}");
        DisplayText();
        GetUserInput();
        CalculateAndDisplay();
    }

    static void Main(string[] args)
    {
        string lang = "ru", level = "medium";
        bool showStats = false, resetStats = false;
        for (int i = 0; i < args.Length; i++)
        {
            string arg = args[i];
            if (arg == "ru" || arg == "en") lang = arg;
            else if (arg == "easy" || arg == "medium" || arg == "hard") level = arg;
            else if (arg == "-s" || arg == "--stats") showStats = true;
            else if (arg == "-r" || arg == "--reset") resetStats = true;
            else if (arg == "-h" || arg == "--help")
            {
                Console.WriteLine("Usage: typing_trainer [ru|en] [easy|medium|hard] [-s] [-r]");
                return;
            }
        }
        if (resetStats)
        {
            foreach (var l in new[]{"ru","en"})
            {
                string f = Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.UserProfile), $".typing_trainer_{l}.json");
                if (File.Exists(f)) File.Delete(f);
            }
            Console.WriteLine("Статистика сброшена.");
            return;
        }
        if (showStats)
        {
            foreach (var l in new[]{"ru","en"})
            {
                string f = Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.UserProfile), $".typing_trainer_{l}.json");
                if (File.Exists(f))
                {
                    try
                    {
                        string json = File.ReadAllText(f);
                        var stats = JsonSerializer.Deserialize<Stats>(json);
                        Console.WriteLine(Colorize($"📊 Статистика для {l}:", "bold"));
                        Console.WriteLine($"  Лучший WPM: {stats.best_wpm:F1}");
                        Console.WriteLine($"  Сыграно игр: {stats.games}");
                        Console.WriteLine($"  Средняя точность: {stats.total_accuracy:F1}%");
                    }
                    catch { Console.WriteLine($"Статистика для {l} повреждена."); }
                }
                else
                {
                    Console.WriteLine($"Статистика для {l} пуста.");
                }
            }
            return;
        }
        TypingTrainer trainer = new TypingTrainer(lang, level);
        trainer.Run();
    }
}

File: typing_trainer.js
// typing_trainer.js
#!/usr/bin/env node
'use strict';

const fs = require('fs');
const path = require('path');
const os = require('os');
const readline = require('readline');

const COLORS = {
    reset: '\x1b[0m',
    green: '\x1b[92m',
    red: '\x1b[91m',
    yellow: '\x1b[93m',
    blue: '\x1b[94m',
    bold: '\x1b[1m'
};

function colorize(text, color) {
    return COLORS[color] + text + COLORS.reset;
}

const WORDS = {
    ru: {
        easy: ['кот','нос','рот','дом','лес','море','вода','рука','нога','мир'],
        medium: ['компьютер','программа','алгоритм','интернет','сервер','сайт','код','данные'],
        hard: ['программирование','искусственный интеллект','коммуникация','информация','технология']
    },
    en: {
        easy: ['cat','dog','sun','run','jump','tree','book','fish','bird','hand'],
        medium: ['computer','program','algorithm','internet','server','website','code','data'],
        hard: ['programming','artificial intelligence','communication','information','technology']
    }
};

class Trainer {
    constructor(lang, level) {
        this.lang = lang;
        this.level = level;
        this.statsFile = path.join(os.homedir(), `.typing_trainer_${lang}.json`);
        this.loadStats();
        this.generateText();
    }

    loadStats() {
        try {
            this.stats = JSON.parse(fs.readFileSync(this.statsFile, 'utf8'));
        } catch {
            this.stats = { best_wpm: 0, games: 0, total_accuracy: 0 };
        }
    }

    saveStats() {
        fs.writeFileSync(this.statsFile, JSON.stringify(this.stats, null, 2));
    }

    generateText() {
        const words = WORDS[this.lang][this.level];
        const count = this.level === 'hard' ? Math.floor(Math.random() * 8) + 8 : Math.floor(Math.random() * 6) + 5;
        let text = '';
        for (let i = 0; i < count; i++) {
            if (i > 0) {
                if (this.level === 'hard' && Math.random() < 0.3) {
                    text += ', ';
                } else {
                    text += ' ';
                }
            }
            const word = words[Math.floor(Math.random() * words.length)];
            text += word;
            if (this.level === 'hard' && (i === count-1 || Math.random() < 0.2)) {
                text += '. ';
            }
        }
        this.text = text.trim();
    }

    displayText() {
        console.log(colorize('\nНапечатайте следующий текст:', 'bold'));
        console.log(colorize(this.text, 'blue'));
        console.log('\nНажмите Enter после завершения ввода. Время начнётся с первой буквы.');
        console.log('Для выхода нажмите Ctrl+C.\n');
    }

    async getUserInput() {
        const rl = readline.createInterface({
            input: process.stdin,
            output: process.stdout
        });
        this.start = Date.now();
        const input = await new Promise(resolve => {
            rl.question('', answer => {
                resolve(answer);
                rl.close();
            });
        });
        this.userInput = input;
        this.end = Date.now();
    }

    calculateAndDisplay() {
        const original = this.text;
        const typed = this.userInput || '';
        const total = original.length;
        const typedLen = typed.length;
        let correct = 0;
        let errors = 0;
        for (let i = 0; i < total; i++) {
            if (i < typedLen && original[i] === typed[i]) {
                correct++;
            } else {
                errors++;
            }
        }
        if (typedLen < total) errors += total - typedLen;
        const elapsed = (this.end - this.start) / 1000;
        const minutes = elapsed / 60;
        const wpm = (correct / 5) / minutes;
        const accuracy = (correct / total) * 100;

        console.log(colorize('\nРезультат:', 'bold'));
        for (let i = 0; i < total; i++) {
            if (i < typedLen && original[i] === typed[i]) {
                process.stdout.write(colorize(original[i], 'green'));
            } else {
                process.stdout.write(colorize(original[i], 'red'));
            }
        }
        console.log();

        console.log(colorize(`Скорость: ${wpm.toFixed(1)} WPM`, 'yellow'));
        console.log(`Точность: ${accuracy.toFixed(1)}%`);
        console.log(`Правильных символов: ${correct}/${total}`);
        console.log(`Ошибок: ${errors}`);
        console.log(`Время: ${elapsed.toFixed(2)} сек`);

        this.stats.games++;
        if (wpm > this.stats.best_wpm) {
            this.stats.best_wpm = wpm;
            console.log(colorize('🏆 Новый рекорд!', 'green'));
        }
        this.stats.total_accuracy = (this.stats.total_accuracy * (this.stats.games-1) + accuracy) / this.stats.games;
        this.saveStats();
        console.log(colorize(`Лучший результат: ${this.stats.best_wpm.toFixed(1)} WPM`, 'blue'));
    }

    async run() {
        console.log(colorize('⌨️  Тренажёр слепой печати', 'bold'));
        console.log(`Язык: ${this.lang}, Уровень: ${this.level}`);
        this.displayText();
        await this.getUserInput();
        this.calculateAndDisplay();
    }
}

async function main() {
    let lang = 'ru', level = 'medium';
    let showStats = false, resetStats = false;
    const args = process.argv.slice(2);
    for (let i = 0; i < args.length; i++) {
        const arg = args[i];
        if (arg === 'ru' || arg === 'en') lang = arg;
        else if (arg === 'easy' || arg === 'medium' || arg === 'hard') level = arg;
        else if (arg === '-s' || arg === '--stats') showStats = true;
        else if (arg === '-r' || arg === '--reset') resetStats = true;
        else if (arg === '-h' || arg === '--help') {
            console.log('Usage: node typing_trainer.js [ru|en] [easy|medium|hard] [-s] [-r]');
            return;
        }
    }
    if (resetStats) {
        ['ru', 'en'].forEach(l => {
            const f = path.join(os.homedir(), `.typing_trainer_${l}.json`);
            if (fs.existsSync(f)) fs.unlinkSync(f);
        });
        console.log('Статистика сброшена.');
        return;
    }
    if (showStats) {
        ['ru', 'en'].forEach(l => {
            const f = path.join(os.homedir(), `.typing_trainer_${l}.json`);
            if (fs.existsSync(f)) {
                const stats = JSON.parse(fs.readFileSync(f, 'utf8'));
                console.log(colorize(`📊 Статистика для ${l}:`, 'bold'));
                console.log(`  Лучший WPM: ${stats.best_wpm.toFixed(1)}`);
                console.log(`  Сыграно игр: ${stats.games}`);
                console.log(`  Средняя точность: ${stats.total_accuracy.toFixed(1)}%`);
            } else {
                console.log(`Статистика для ${l} пуста.`);
            }
        });
        return;
    }
    const trainer = new Trainer(lang, level);
    await trainer.run();
}

main().catch(console.error);
